Throw when `layer(…)` in `@import` is in the incorrect spot

RobinMalfait opened the pull request:
This PR throws an error when we notice that an `layer(…)` in an `@import` or `@media` is incorrect.

This hints the user to ensure that `layer(…)` in an `@import` should be the first condition. In case of an `@media`, it should be an `@layer …` instead.

RobinMalfait committed:
throw when `layer(…)` in import is found

RobinMalfait committed:
update changelog

adamwathan commented on CHANGELOG.md:
```diff
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - Use configured `--letter-spacing` values for custom font size utilities ([#15099](https://github.com/tailwindlabs/tailwindcss/pull/15099))
 - Ensure `space-x/y-*` and `divide-x/y-*` with variants can undo `space-x/y-reverse` and `divide-x/y-reverse` ([#15094](https://github.com/tailwindlabs/tailwindcss/pull/15094))
+- Throw when `layer(…)` in `@import` is in the incorrect spot ([#15109](https://github.com/tailwindlabs/tailwindcss/pull/15109))
```
```suggestion
- Error when `layer(…)` in `@import` is not first in the list of functions/conditions ([#15109](https://github.com/tailwindlabs/tailwindcss/pull/15109))
```

adamwathan commented on packages/tailwindcss/src/index.ts:
```diff
@@ -337,6 +337,77 @@ async function parseCss(
+            expected.splice(2, 0, ...expected.splice(layer, 1))
+
+            let errorMessage = [
+              'A `layer(…)` in an `@import` should come first:',
```
```suggestion
              '`layer(…)` in an `@import` should come before any other functions or conditions:',
```

adamwathan commented on packages/tailwindcss/src/index.ts:
```diff
@@ -337,6 +337,77 @@ async function parseCss(
+            }
+
+            let errorMessage = [
+              'A `@media layer(…)` is invalid, did you mean to use `@layer`?',
```
When would this happen?

RobinMalfait replied:
If you have written that manually or used a pre-processor that doesn't handle it properly. But I think we can just get rid of this because we don't do validation like this anywhere else.

RobinMalfait committed:
Update packages/tailwindcss/src/index.ts

Co-authored-by: Adam Wathan <adam.wathan@gmail.com>

RobinMalfait committed:
revert `@import layer(…)` handling main CSS handler

RobinMalfait committed:
forward thrown errors from `parseImportParams(…)`

RobinMalfait committed:
ignore all errors

If any of the errors occur, we can bail and keep the `@import` in the
final CSS.

If something went wrong, the browser will let you know that something is
missing.

philipp-spiess commented on CHANGELOG.md:
```diff
@@ -7,7 +7,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [Unreleased]
 
-- Nothing yet!
+### Fixed
```
```suggestion
### Changed
```

The error alone doesn't really fix anything heh

philipp-spiess commented on CHANGELOG.md:
```diff
@@ -7,7 +7,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
-- Nothing yet!
+### Fixed
+
+- Error when `layer(…)` in `@import` is not first in the list of functions/conditions ([#15109](https://github.com/tailwindlabs/tailwindcss/pull/15109))
```
```suggestion
- Error when `layer(…)` as an argument to the `@import` rule is not on first place ([#15109](https://github.com/tailwindlabs/tailwindcss/pull/15109))
```

Not sure if this is really clearer though :/  Take this as an idea only lol

RobinMalfait replied:
This was already an updated version from @adamwathan haha

See: https://github.com/tailwindlabs/tailwindcss/pull/15109#discussion_r1854128092

philipp-spiess replied:
😬 in that case, nothing to see here

RobinMalfait committed:
Update CHANGELOG.md

Co-authored-by: Philipp Spiess <hello@philippspiess.com>

RobinMalfait committed:
Merge branch 'next' into feat/warn-when-invalid-layer-is-found
